build-llvm.py

```diff
@@ -0,0 +1,435 @@
+        # https://clang.llvm.org/docs/AutomaticReferenceCounting.html
+        'CLANG_ENABLE_ARCMT': 'OFF',
+        # We don't (currently) use the static analyzer and it saves cycles
+        # according to Chromium OS:
+        # https://crrev.com/44702077cc9b5185fc21e99485ee4f0507722f82
+        'CLANG_ENABLE_STATIC_ANALYZER': 'OFF',
+        # We don't use the plugin system and it will remove unused symbols:
+        # https://crbug.com/917404
+        'CLANG_PLUGIN_SUPPORT': 'OFF',
+        # The C compiler to use
+        'CMAKE_C_COMPILER': env_vars.cc,
+        # The C++ compiler to use
+        'CMAKE_CXX_COMPILER': env_vars.cxx,
+        # Where the toolchain should be installed
+        'CMAKE_INSTALL_PREFIX': install_folder.as_posix(),
+        # For LLVMgold.so, which is used for LTO with ld.gold
+        'LLVM_BINUTILS_INCDIR': root_folder.joinpath(utils.current_binutils(), "include").as_posix(),
+        # The projects to build
+        'LLVM_ENABLE_PROJECTS': projects,
+        # Don't build bindings; they are for other languages that the kernel does not use
+        'LLVM_ENABLE_BINDINGS': 'OFF',
+        # Don't build Ocaml documentation
+        'LLVM_ENABLE_OCAMLDOC': 'OFF',
+        # Removes system dependency on terminfo and almost every major clang provider turns this off
+        'LLVM_ENABLE_TERMINFO': 'OFF',
+        # Don't build clang-tools-extras to cut down on build targets (about 400 files or so)
+        'LLVM_EXTERNAL_CLANG_TOOLS_EXTRA_SOURCE_DIR': '',
+        # Don't include documentation build targets because it is available on the web
+        'LLVM_INCLUDE_DOCS': 'OFF',
+        # Don't include example build targets to save on cmake cycles
+        'LLVM_INCLUDE_EXAMPLES': 'OFF',
+        # The architectures to build backends for
+        'LLVM_TARGETS_TO_BUILD': targets
+    }
+    # yapf: enable
+
+    # If a debug build was requested
+    if debug:
+        defines['CMAKE_BUILD_TYPE'] = 'Debug'
+        defines['CMAKE_C_FLAGS'] = '-march=native -mtune=native'
+        defines['CMAKE_CXX_FLAGS'] = '-march=native -mtune=native'
+        defines['LLVM_BUILD_TESTS'] = 'ON'
+    # If a release build was requested
+    else:
+        defines['CMAKE_BUILD_TYPE'] = 'Release'
+        defines['CMAKE_C_FLAGS'] = '-O2 -march=native -mtune=native'
+        defines['CMAKE_CXX_FLAGS'] = '-O2 -march=native -mtune=native'
+        defines['LLVM_INCLUDE_TESTS'] = 'OFF'
+        defines['LLVM_ENABLE_WARNINGS'] = 'OFF'
+
+    # Don't build libfuzzer when compiler-rt is enabled, it invokes cmake again and we don't use it
+    if "compiler-rt" in projects:
+        defines['COMPILER_RT_BUILD_LIBFUZZER'] = 'OFF'
+
+    # Use ccache if it is available for faster incremental builds
+    if shutil.which("ccache") is not None:
+        defines['LLVM_CCACHE_BUILD'] = 'ON'
+
+    # If we found a linker, we should use it
+    if env_vars.ld is not None:
+        defines['LLVM_USE_LINKER'] = env_vars.ld
```

Comment:
Does this work? I just tried this manually, and it seems that the check this does doesn't work when building clang w/ clang (ie when using `-DCMAKE_CXX_COMPILER` which seems like a deficiency with [`check_cxx_source_compiles`](https://cmake.org/cmake/help/latest/module/CheckCXXSourceCompiles.html) not passing along cmake flags to [`try_compile`](https://cmake.org/cmake/help/latest/command/try_compile.html#command:try_compile).

Reply:
Hmmm, what does the print out of `CC`, `CXX`, and `LD` show at the beginning of the script? It works with all of the distributions I have checked in Docker.

Reply:
Right, but you're probably *not* setting `-DCMAKE_C_COMPILER` and `-DCMAKE_CXX_COMPILER`.  Locally, I set these to Android's prebuilts in order to build clang with clang.  I suppose this workflow is "unsupported" in cmake...

Reply:
Those values are defined in this script on lines 318 and 320.

Reply:
ah, thanks to @pirama-arumuga-nainar for pointing out `-DLLVM_ENABLE_LLD=ON` is a better approach.

Reply:
Only in the case of `ld.lld`. The user may be building with GCC which won't support that option or be using `ld.gold`.

`-DLLVM_USE_LINKER` should work no problem with `clang`, I'm curious what the fail looks like on your end.